### PR TITLE
Register `EnvironmentType`s using their classes

### PR DESCRIPTION
**Pull request by serhii-lekariev**
This PR introduces several changes to `Environment` and `EnvironmentType`s.

- `Tests` and `Production` now have `package-private` constructors;
- `Environment` now caches the class of the active `EnvironmentType` instead of an instance;
- `Environment` now exposes an `@Internal` method to register an `EnvironmentType` by class. Classes to be registered using this method must have a parameterless constructor. This limitation is caused by the fact that we have to instantiate the specified `EnvironmentType`.

This leads to a signature change: `public EnvironmentType type()` -> `public Class<? extends EnvironmentType> type()`.
It also allows to add a `setTo(Class<? extends EnvironmentType>)` method, which is now necessary, as users can no longer `setTo(new Tests())`, as `Tests` now has a package-private constructor.

#### `Invokables`

This PR also extracts some of the reflection-related utilities into an `Invokables` class.

Users of `Methods` may found the utilities there.

### CODE REVIEW

**Comment by serhii-lekariev** [comment]
@armiol, PTAL again.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL again.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL again.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

File: base/src/main/java/io/spine/base/Environment.java [hunk 1]
```diff
@@ -159,6 +192,27 @@ public Environment register(EnvironmentType environmentType) {
         return this;
     }
 
+    /**
+     * Remembers the specified environment type, allowing {@linkplain #is(Class) to
+     * determine whether it's enabled} later.
+     *
+     * <p>The specified {@code type} must have a parameterless constructor.
+     * {@linkplain EnvironmentType#enabled() activity} of the specified environment type is going
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:200]: `activity` should start from the capital letter. But I am not sure I understand this sentence:
```
activity of the specified environment type is going to be checked against an instance created by invoking the parameterless constructor.
```
Please rewrite into two or more _simple_ sentences.

File: base/src/main/java/io/spine/base/Environment.java [hunk 2]
```diff
@@ -104,6 +107,36 @@
...
+ *     // is `true`.
+ *     assertThat(environment.is(AwsLambda.class)).isTrue();
+ *
+ *     environment.reset();
+ *
+ *     // When `reset` explicitly, cached value is erased.
+ *     assertThat(environment.is(AwsLambda.class)).isFalse();
+ * </pre>
+ *
+ * <p>{@linkplain #setTo(EnvironmentType) explicitly setting} the environment type overrides
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:137]:
```
When set {@linkplain ... explicitly}, ...
```

Notice the capital letter.

File: base/src/main/java/io/spine/base/Environment.java [hunk 3]
```diff
@@ -104,6 +107,36 @@
  * }
  * </pre>
  *
+ * <h1>Caching</h1>
+ * <p>{@code Environment} caches the {@code EnvironmentType} once its calculated.
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:111]: For the sake of readability of this `.java` file, please add an empty line above.

File: base/src/main/java/io/spine/base/Environment.java [hunk 4]
```diff
@@ -104,6 +107,36 @@
...
+ * This means that if one environment type has been found to be active, its instance is saved.
+ * If later it becomes logically inactive, e.g. the environment variable that's used to check the
+ * environment type changes, {@code Environment} is still going to return the cached value, unless
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:114]: Don't use complex sentences.

```
... going to return the cached value. To overwrite the value (do something). Also, the value may be {@linkplain ... reset}.
```

**Comment by serhii-lekariev** [comment]
@armiol, PTAL again.

File: base/src/test/java/io/spine/base/EnvironmentTest.java [hunk 5]
```diff
@@ -217,32 +232,63 @@ void determineUsingType() {
...
+        assertThat(environment.is(Staging.class)).isTrue();
     }
 
+    @Test
+    @DisplayName("cache the `Tests` environment type")
+    void cacheTests() throws InterruptedException {
+        AtomicBoolean envCached = new AtomicBoolean(false);
+        assertThat(environment.is(Tests.class));
+        Thread thread = new Thread(() -> {
+            /*
```
Comment (armiol) [inline base/src/test/java/io/spine/base/EnvironmentTest.java:275]: Please convert this to the method-level comment.

Don't use inline comments except for emergency cases.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

File: base/src/main/java/io/spine/base/Environment.java [hunk 6]
```diff
@@ -299,19 +299,26 @@ public void restoreFrom(Environment copy) {
 
     /**
      * Sets the current environment type to {@code type.getClass()}. Overrides the current value.
+     *
+     * <p>Calls {@link #register(EnvironmentType)} internally.
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:303]:
```
If the supplied type was not {@linkplain #register(EnvironmentType) registered} previously, it is registered.
```

File: base/src/main/java/io/spine/base/Environment.java [hunk 7]
```diff
@@ -299,19 +299,26 @@ public void restoreFrom(Environment copy) {
...
-        this.currentEnvType = checkNotNull(type).getClass();
+        checkNotNull(type);
+        register(type);
+        this.currentEnvType = type.getClass();
     }
 
     /**
      * Sets the current environment type to the specified one. Overrides the current value.
+     *
+     * <p>Calls {@link #register(Class)} internally.
```
Comment (armiol) [inline base/src/main/java/io/spine/base/Environment.java:315]: Same as above.

**Pull request merged**